### PR TITLE
Add redis firewallRules to swagger specs w/ examples. (+minor descrip…

### PR DESCRIPTION
**Pull request by TimLovellSmith**
…tive changes to existing APIs).
(Reissue of PR #977 to get around CI issues...)

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [x] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

### CODE REVIEW

**Comment by TimLovellSmith** [comment]
(Please assign this to @veronicagg who had #977!)

**Comment by TimLovellSmith** [comment]
@veronicagg It looks like there is some problem with the CI installing/running autorest for node.js?

**Comment by TimLovellSmith** [comment]
@veronicagg 
I couldn't figure this message from https://travis-ci.org/Azure/azure-rest-api-specs/jobs/207499990 out

7) Azure swagger model validation using x-ms-examples and examples in spec /home/travis/build/Azure/azure-rest-api-specs/arm-redis/2016-04-01/swagger/redis.json should have valid examples.:

3360     ReferenceError: error is not defined

3361      at SpecValidator.validateRequest (node_modules/openapi-validation-tools/lib/specValidator.js:567:79)

3362      at SpecValidator.validateXmsExamples (node_modules/openapi-validation-tools/lib/specValidator.js:470:60)

3363      at SpecValidator.validateOperation (node_modules/openapi-validation-tools/lib/specValidator.js:412:10)

3364      at SpecValidator.validateOperations (node_modules/openapi-validation-tools/lib/specValidator.js:446:12)

3365      at node_modules/openapi-validation-tools/validate.js:86:15

**Comment by TimLovellSmith** [comment]
Still getting a cryptic build error after fixing parameter names.
Shouldn't need to be tagged cla-required for me...

**Comment by amarzavery** [comment]
The github CLA bot does not like you :(. It does the same with me many times :(

File: arm-redis/2016-04-01/swagger/redis.json [hunk 1]
```diff
@@ -545,6 +545,213 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cache/Redis/{cacheName}/firewallRules/": {
```
Comment (amarzavery) [inline arm-redis/2016-04-01/swagger/redis.json]: I think there is an extra slash in the end that shouldn't be there, correct ?
"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cache/Redis/{cacheName}/firewallRules**/**":

Comment (TimLovellSmith) [inline arm-redis/2016-04-01/swagger/redis.json]: @amarzavery thanks for spotting that!

**Comment by TimLovellSmith** [comment]
@amarzavery What should I do about the CLA issue? Node SDK generation failure is opaque. I also see a lot of linter errors which seem to be extraneous to the content of the specific PR, I think they are caused by improvements to the linter - are any of those going to be blockers for this?

**Comment by TimLovellSmith** [comment]
@veronicagg any advice on how to deal with the linter errors here?

**Review by veronicagg — CHANGES_REQUESTED** [review]
@TimLovellSmith I made some comments inline related to linter errors. And here are some comments for the others: 
Is there an operations API you can expose for the service? This error https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208317608#L979 comes from a request from the ARM team to have an operation in the spec that lists all the available operations, like: https://github.com/Azure/azure-rest-api-specs/blob/master/arm-cdn/2016-10-02/swagger/cdn.json#L1578 

I'm following up on errors with code "ArmResourcePropertiesBag", as I believe the rule may not be appropriate for those cases. (https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208317608#L1001)

Even though, this issue is not caused by this PR, please take a look at https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208317608#L1090

**Comment by TimLovellSmith** [comment]
@veronicagg What contexts will humans be exposed to the operation name FirewallRules_List/firewallrulesListByRedis in? Do operation names have to be globally unique for any reason (e.g. using as programmatic ID)? The latter is not an intuitive name, so I'd prefer FirewallRules_List if possible, I just want to make sure there's no extra constraints...

File: arm-redis/2016-04-01/swagger/redis.json [hunk 2]
```diff
@@ -545,6 +545,213 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Cache/Redis/{cacheName}/firewallRules": {
+      "get": {
+        "tags": [
+          "Redis",
+          "FirewallRules"
+        ],
+        "operationId": "RedisFirewallRules_List",
```
Comment (TimLovellSmith) [inline arm-redis/2016-04-01/swagger/redis.json]: @veronicagg The naming guidance seems a little arbitrary. E.g. We call the create operation RedisFirewallRule_CreateOrUpdate and there's no error for that? (Or I overlooked it)

**Comment by TimLovellSmith** [comment]
@veronicagg I've made the change with 'default' on it - I do still have a worry here that this might have been overvalidation - as it might end up making the actual generated sdk client more confusing to have to specify an enumeration with a single value... but we can wait to preview what sdk changes would result.

**Comment by TimLovellSmith** [comment]
@veronicagg 
Re: I've made the change with patchSchedules 'default' as enum" - now it seems to just fail model validation even worse... I think I made the required changes to examples here though.

```
{ code: 'REQUEST_VALIDATION_ERROR',
  message: 'Found errors in validating the request for x-ms-example "RedisCachePatchSchedulesGet" in operation "PatchSchedules_Get".',
  innerErrors: 
   [ { code: 'REQUIRED_PARAMETER_EXAMPLE_NOT_FOUND',
       message: 'In operation "PatchSchedules_Get", parameter name is required in the swagger spec but is not present in the provided example parameter values.' } ] }
{ AssertionError: swagger "arm-redis/2016-04-01/swagger/redis.json" contains model validation errors.
    at /home/travis/build/Azure/azure-rest-api-specs/test/model.js:21:14
  name: 'AssertionError',
  actual: false,
  expected: true,
  operator: '==',
  message: 'swagger "arm-redis/2016-04-01/swagger/redis.json" contains model validation errors.',
  generatedMessage: false }
    1) arm-redis/2016-04-01/swagger/redis.json should have valid examples.
```

**Comment by TimLovellSmith** [comment]
@veronicagg No, my bad, I didn't update parameter set for all METHODs.

**Comment by veronicagg** [comment]
@TimLovellSmith for your fist question "FirewallRules_List" looks good, the operation id is used in the method name in the SDK.
Regarding the parameter/enum question, see https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-extensions.md#single-value-enum-as-a-constant

**Comment by TimLovellSmith** [comment]
@veronicagg Thanks for the answers! I'll change it to FirewallRules_List then, and bring back the patchSchedules/default change.

**Comment by TimLovellSmith** [comment]
@veronicagg So I think I've fixed all the important feedbacks and linter errors. However, node SDK build job still seems to be failing in TravisCI - and outputting usage instructions instead of an actual error!? Could you please check on that?

**Comment by veronicagg** [comment]
@TimLovellSmith we're working on fixing the CI failure that you're seeing, we expect to have that working today. 
Changes look good, are you planning on fixing this one too?

- Is there an "operations API" you can expose for the service? This error https://travis-ci.org/Azure/azure-rest-api-specs/jobs/208317608#L979 comes from a request from the ARM team to have an operation in the spec that lists all the available operations, like: https://github.com/Azure/azure-rest-api-specs/blob/master/arm-cdn/2016-10-02/swagger/cdn.json#L1578

Thanks!

**Comment by TimLovellSmith** [comment]
@veronicagg No, I was not planning to address that as part of this PR. Is there an easy way to define the operations API without having to define the entire operation operations body shape in our redis.json? It's rather a lot of declarations to add, and I would think usually every RP implements the same contract.

**Comment by veronicagg** [comment]
@TimLovellSmith There seem to still be issues with CI, not just yours, so I've pinged someone else on that. I'll restart this job in case it helps anyway. 
Regarding the OperationAPI: does the way that [CDN](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-cdn/2016-10-02/swagger/cdn.json#L1578) defines it not work well for yours? What type of "easier" way would you be looking for in this case?
One more comment, after chatting with @ravbhatnagar a bit more, our validation for ProvidersPathValidation is a bit to strict and defining a parameter as a constant is not required, so you can leave it as is, or revert back to having the value in the path at https://github.com/Azure/azure-rest-api-specs/pull/989/files#diff-785771b0c72c01e60191a5d798d6d508R755; we'll be updating our validation tool to account for this soon.

**Comment by TimLovellSmith** [comment]
@veronicagg Re "Operations" api, I think I was hoping for something like "Resource" such that every RP definition can link to one common definition, since they should all follow the same basic contract. Although yes, "Resource" is in our spec too...

**Comment by veronicagg** [comment]
@TimLovellSmith ah, yes, I believe you're asking for https://github.com/Azure/azure-rest-api-specs-pr/pull/40, it's coming, at which point you'll be able to point to the common definition.

**Comment by TimLovellSmith** [comment]
@veronicagg I had a look at pull/40 but I don't see Operations anywhere in that diff? It feels nitpicky that we are discussing linter errors Operations API as part of this PR as it is really orthogonal to the changes proposed. As I looked into implementing CDN's declared contract for Operations does not match our RP's actual behavior... in investigating why, it appears that CDN's contract does not match the specification either! [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#available-ops-id]

**Comment by TimLovellSmith** [comment]
@veronicagg I've added the Operations API... its definition slightly varies from CDN's.

File: arm-redis/2016-04-01/swagger/redis.json [hunk 3]
```diff
@@ -1408,6 +1433,53 @@
...
+            "description": {
+              "description": "Friendly name of the operation",
+              "type": "string"
+            }
+          }
+        }
+      }
+    },
+    "OperationListResult": {
+     "description": "Result of the request to list CDN operations. It contains a list of operations and a URL link to get the next set of results.",
```
Comment (veronicagg) [inline arm-redis/2016-04-01/swagger/redis.json]: please update description

File: arm-redis/2016-04-01/swagger/redis.json [hunk 4]
```diff
@@ -1408,6 +1433,53 @@
...
+      "properties": {
+        "value": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/Operation"
+          },
+          "description": "List of CDN operations supported by the CDN resource provider."
```
Comment (veronicagg) [inline arm-redis/2016-04-01/swagger/redis.json]: please update description

**Review by veronicagg — CHANGES_REQUESTED** [review]
Minor comment regarding descriptions. I agree this change is orthogonal with the proposed change, so in my previous comments I was mostly asking if you were planning on addressing it now or not. What's blocking this PR at the moment, is that we're having issues with CI, I've restarted the job a couple of times, there seems to still be failures, which I'm trying to understand.

**Comment by veronicagg** [comment]
@ravbhatnagar do you have any thoughts on the question above regarding OperationsAPI?: "@veronicagg I had a look at pull/40 but I don't see Operations anywhere in that diff? It feels nitpicky that we are discussing linter errors Operations API as part of this PR as it is really orthogonal to the changes proposed. Edit... I suck at reading swagger" thanks!

**Comment by TimLovellSmith** [comment]
@veronicagg Woohoo, checks passed!

**Review by veronicagg — APPROVED** [review]

**Comment by TimLovellSmith** [comment]
@veronicagg Thank you!

**Pull request merged**